crates/storage/src/snapshot.rs

```diff
@@ -238,104 +280,150 @@ impl StateRead for Snapshot {
     }
 
+    /// Returns a stream of all key-value pairs with the given prefix, from nonverifiable storage.
     fn nonverifiable_prefix_raw(&self, prefix: &[u8]) -> Self::NonconsensusPrefixRawStream {
         let span = Span::current();
-        let self2 = self.clone();
+        let rocksdb_snapshot = self.0.snapshot.clone();
+        let db = self.0.db.clone();
+
+        let (truncated_prefix, config) = self.0.multistore_cache.config.match_prefix_bytes(prefix);
+        tracing::debug!(substore_key = ?truncated_prefix,  substore_prefix = config.prefix, prefix_supplied = ?prefix, "matched prefix, fetching substore");
+        let version = self
+            .substore_version(&config)
+            .expect("the substore exists and has been initialized");
+
+        let substore = store::substore::SubstoreSnapshot {
+            config,
+            rocksdb_snapshot,
+            version,
+            db,
+        };
 
         let mut options = rocksdb::ReadOptions::default();
-        options.set_iterate_range(rocksdb::PrefixRange(prefix));
+        options.set_iterate_range(rocksdb::PrefixRange(truncated_prefix));
         let mode = rocksdb::IteratorMode::Start;
 
-        let (tx, rx) = mpsc::channel(10);
+        let (tx_prefix_query, rx_prefix_query) = mpsc::channel(10);
 
-        // Here we're operating on the nonverifiable data, which is a raw k/v store,
-        // so we just iterate over the keys.
         tokio::task::Builder::new()
             .name("Snapshot::nonverifiable_prefix_raw")
             .spawn_blocking(move || {
                 span.in_scope(|| {
-                    let keys_cf = self2
-                        .0
-                        .db
-                        .cf_handle("nonverifiable")
-                        .expect("nonverifiable column family not found");
-                    let iter = self2.0.snapshot.iterator_cf_opt(keys_cf, options, mode);
+                    let cf_nonverifiable = substore.config.cf_nonverifiable(&substore.db);
+                    let iter =
+                        substore
+                            .rocksdb_snapshot
+                            .iterator_cf_opt(cf_nonverifiable, options, mode);
                     for i in iter {
                         let (key, value) = i?;
-                        tx.blocking_send(Ok((key.into(), value.into())))?;
+                        tx_prefix_query.blocking_send(Ok((key.into(), value.into())))?;
                     }
                     anyhow::Ok(())
                 })
             })
             .expect("should be able to spawn_blocking");
 
-        tokio_stream::wrappers::ReceiverStream::new(rx)
+        tokio_stream::wrappers::ReceiverStream::new(rx_prefix_query)
     }
 
+    /// Returns a stream of all key-value pairs with the given prefix, and range
+    /// from nonverifiable storage.
+    /// TODO(erwan): For now this method only supports range queries over the main store.
```

Comment:
It's further afield, but one possible future use case could be to put all of the DEX indexes in a substore, so that DEX queries would operate only on DEX data. But I don't think this is currently the lowest hanging fruit for performance there.

Reply:
Good point, that would be neat. Prefix queries work with substores, but the more general range queries don't yet. There isn't a hard blocker, but it requires massaging the API. It's tracked in [#3327](https://github.com/penumbra-zone/penumbra/issues/3327).